src/isql/isql.epp

```diff
@@ -1822,63 +1822,68 @@ bool ISQL_printNumericType(const char* fieldName, const int fieldType, const int
 		return false;
 	}
 
-	bool precision_known = false;
-
-	if (isqlGlob.major_ods >= ODS_VERSION10 &&
-		(fieldType == SMALLINT || fieldType == INTEGER || fieldType == BIGINT))
+	switch (fieldType)
 	{
+	case SMALLINT:
+	case INTEGER:
+	case BIGINT:
 		// Handle Integral subtypes NUMERIC and DECIMAL
 		// We are ODS >= 10 and could be any Dialect
-		FOR FLD1 IN RDB$FIELDS WITH
-			FLD1.RDB$FIELD_NAME EQ fieldName
-			AND NOT FLD1.RDB$FIELD_PRECISION EQ 0
 
-			// We are Dialect >=3 since FIELD_PRECISION is non-NULL
-			if (FLD1.RDB$FIELD_SUB_TYPE > 0 && FLD1.RDB$FIELD_SUB_TYPE <= MAX_INTSUBTYPES)
+		// We are Dialect >=3 since FIELD_PRECISION is non-NULL
+		if (isqlGlob.major_ods >= ODS_VERSION10 && fieldPrecision != 0 && fieldSubType > 0 && fieldSubType <= MAX_INTSUBTYPES)
+		{
+			isqlGlob.printf("%s(%d, %d)",
+				Integral_subtypes[fieldSubType],
+				fieldPrecision,
+				-fieldScale);
+		}
+		else
+		{
+			if (fieldScale < 0)
 			{
-				isqlGlob.printf("%s(%d, %d)",
-						 Integral_subtypes[FLD1.RDB$FIELD_SUB_TYPE],
-						 FLD1.RDB$FIELD_PRECISION,
-						 -FLD1.RDB$FIELD_SCALE);
-				precision_known = true;
+				// Take a stab at numerics and decimals
+				switch (fieldType)
+				{
+				case SMALLINT:
+					isqlGlob.printf("NUMERIC(4, %d)", -fieldScale);
+					break;
+				case INTEGER:
+					isqlGlob.printf("NUMERIC(9, %d)", -fieldScale);
+					break;
+				case BIGINT:
+					isqlGlob.printf("NUMERIC(18, %d)", -fieldScale);
+					break;
+				case DOUBLE_PRECISION:
+					isqlGlob.printf("NUMERIC(15, %d)", -fieldScale);
+					break;
+				}
 			}
-		END_FOR
-		ON_ERROR
-			ISQL_errmsg (fbStatus);
-			return false;
-		END_ERROR;
-	}
-
-	if (!precision_known)
-	{
-		if (fieldScale < 0)
-		{
-			// Take a stab at numerics and decimals
-			switch (fieldType)
+			else
 			{
-			case SMALLINT:
-				isqlGlob.printf("NUMERIC(4, %d)", -fieldScale);
-				break;
-			case INTEGER:
-				isqlGlob.printf("NUMERIC(9, %d)", -fieldScale);
-				break;
-			case BIGINT:
-				isqlGlob.printf("NUMERIC(18, %d)", -fieldScale);
-				break;
-			case DOUBLE_PRECISION:
-				isqlGlob.printf("NUMERIC(15, %d)", -fieldScale);
 				break;
-			default:
-				isqlGlob.printf("%s", Column_types[i].type_name);
 			}
 		}
-		else
+		return true;
+	case T_CHAR:
```

Comment:
I keep patch as small as possible. Refactoring or cleanup unrelated things will cause rejection of the patch.